server/topic.go

```diff
@@ -1125,32 +1140,53 @@ func (t *Topic) thisUserSub(h *Hub, sess *Session, asUid types.Uid, asLvl auth.L
 			}
 		}
 
-		// Undelete
+		// Undelete.
 		userData.deleted = false
 
 		if isNullValue(private) {
 			private = nil
 		}
 		userData.private = private
 
-		// Add subscription to database
-		sub := &types.Subscription{
-			User:      asUid.String(),
-			Topic:     t.name,
-			ModeWant:  userData.modeWant,
-			ModeGiven: userData.modeGiven,
-			Private:   userData.private,
-		}
+		tname := t.name
+		var sub *types.Subscription
+		if isChanSub {
+			// Check if user is already subscribed.
+			var err error
+			sub, err = store.Subs.Get(pkt.Original, asUid)
+			if err != nil {
+				sess.queueOut(ErrUnknown(pkt.Id, toriginal, now))
+				return changed, err
+			}
+			tname = pkt.Original
+			userData.modeWant = types.ModeCChn
```

Comment:
Move the permission changes block to line 1141 (else clause) for better readability?

Reply:
done